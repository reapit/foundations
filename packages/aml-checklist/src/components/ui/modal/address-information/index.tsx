import AddressInformation from './address-information'
export default AddressInformation
