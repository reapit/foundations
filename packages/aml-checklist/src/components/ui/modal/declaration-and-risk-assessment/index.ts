import DeclarationAndRiskAssessment from './declaration-and-risk-assessment'
export default DeclarationAndRiskAssessment
