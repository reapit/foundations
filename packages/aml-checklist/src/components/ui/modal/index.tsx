import Modal from './modal'

export default Modal

export const STEPS = {
  PROFILE: 'Personal Details',
  PRIMARY_IDENTIFICATION: 'Primary ID',
  SECONDARY_IDENTIFICATION: 'Secondary ID',
  ADDRESS_INFORMATION: 'Address Information',
  REPORT: 'Report',
  DECLARATION_RISK_MANAGEMENT: 'Declaration Risk Management',
  EXPERIAN: 'Experian',
  PEP_SEARCH: 'PEP Search',
}
