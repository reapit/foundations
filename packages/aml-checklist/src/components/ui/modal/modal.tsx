import React from 'react'
import { Modal } from '@reapit/elements-legacy'
import Profile from './profile'
import Report from './report'
import AddressInformation from './address-information'
import PrimaryIdentification from './primary-identification'
import SecondaryIdentification from './secondary-identification'
import DeclarationAndRiskAssessment from './declaration-and-risk-assessment'
import PepSearch from './pep-search'
import UpdateStatus from './update-status'
import UpdateStatusSuccess from './update-status-success'
import { withRouter, RouteComponentProps } from 'react-router'

export const STEPS = {
  PROFILE: 'Personal Details',
  PRIMARY_IDENTIFICATION: 'Primary ID',
  SECONDARY_IDENTIFICATION: 'Secondary ID',
  ADDRESS_INFORMATION: 'Address Information',
  REPORT: 'Report',
  DECLARATION_RISK_MANAGEMENT: 'Declaration Risk Management',
  EXPERIAN: 'Experian',
  PEP_SEARCH: 'PEP Search',
}

export const ID_STATUS = {
  UPDATE: 'Update Status',
  SUCCESS: 'Update Status Success',
}

export type ProfileModalProps = {
  modalContentType: string
  visible: boolean
  id: string
  afterClose: () => void
} & RouteComponentProps

export const renderContent = ({ modalContentType, history }) => {
  switch (modalContentType) {
    case STEPS.PROFILE:
      return <Profile />

    case STEPS.PRIMARY_IDENTIFICATION:
      return <PrimaryIdentification />

    case STEPS.SECONDARY_IDENTIFICATION:
      return <SecondaryIdentification />

    case STEPS.ADDRESS_INFORMATION:
      return <AddressInformation />

    case STEPS.REPORT:
      return <Report />

    case STEPS.PEP_SEARCH:
      return <PepSearch />

    case STEPS.DECLARATION_RISK_MANAGEMENT:
      return <DeclarationAndRiskAssessment />

    case ID_STATUS.UPDATE:
      return <UpdateStatus />

    case ID_STATUS.SUCCESS:
      return <UpdateStatusSuccess history={history} />
    default:
      return null
  }
}

export const ProfileModal: React.FC<ProfileModalProps> = ({
  visible,
  afterClose,
  modalContentType = STEPS.PROFILE,
  history,
}) => {
  return (
    <Modal title={modalContentType} visible={visible} size="medium" afterClose={afterClose}>
      <div>{renderContent({ modalContentType, history })}</div>
    </Modal>
  )
}

export default withRouter(ProfileModal)
