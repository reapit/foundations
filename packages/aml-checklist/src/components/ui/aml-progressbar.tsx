import React from 'react'
import {
  ProgressBar,
  H3,
  LevelRight,
  LevelItem,
  LevelLeft,
  Level,
  Button,
  AcLink,
  EntityType,
  SubTitleH5,
  FadeIn,
  Section,
} from '@reapit/elements'
import { SectionsStatus } from '@/reducers/checklist-detail'
import { IdentityCheckModel, ContactModel } from '@reapit/foundations-ts-definitions'
import { Dispatch } from 'redux'
import { checklistDetailShowModal } from '@/actions/checklist-detail'
import { ReduxState } from '@/types/core'
import { connect } from 'react-redux'
import {
  selectCheckListDetailStatus,
  selectCheckListDetailContact,
  selectCheckListDetailIdCheck,
} from '@/selectors/checklist-detail'
import { ID_STATUS } from './modal/modal'
import { useReapitConnect } from '@reapit/connect-session'
import { reapitConnectBrowserSession } from '@/core/connect-session'

export type AMLProgressBarProps = AMLProgressBarMappedActions & AMLProgressBarMappedProps

// TODO: will replace when get data from BE
export const calculateProgress = (status: SectionsStatus) => {
  const count = Object.keys(status).length
  const completedCount = Object.keys(status).filter(key => status[key]).length
  return { percentage: Math.floor((completedCount / count) * 100), completed: completedCount, total: count }
}

export const AMLProgressBar: React.FC<AMLProgressBarProps> = ({ contact, idCheck, status, showModal }) => {
  const progress = React.useMemo(() => calculateProgress(status), [status])
  const { connectIsDesktop } = useReapitConnect(reapitConnectBrowserSession)

  const { id, title, forename, surname } = contact || {}
  const name = `${title || ''} ${forename || ''} ${surname || ''}`.trim()

  return (
    <FadeIn>
      <Section>
        <Level>
          <LevelLeft>
            <LevelItem>
              <H3>
                <AcLink
                  dynamicLinkParams={{
                    appMode: connectIsDesktop ? 'DESKTOP' : 'WEB',
                    entityType: EntityType.CONTACT,
                    entityCode: id,
                  }}
                >
                  {name}
                </AcLink>
              </H3>
              <SubTitleH5>{idCheck && idCheck.status && `Status: ${idCheck.status.toUpperCase()}`}</SubTitleH5>
            </LevelItem>
          </LevelLeft>
          <LevelRight>
            <LevelItem>
              <Button disabled={!idCheck} type="button" variant="primary" onClick={() => showModal(ID_STATUS.UPDATE)}>
                Update Status
              </Button>
            </LevelItem>
          </LevelRight>
        </Level>
        <ProgressBar percentage={progress.percentage} />
        <div className="has-text-right">
          {progress.completed}/{progress.total} <span>Completed</span>
        </div>
      </Section>
    </FadeIn>
  )
}

export interface AMLProgressBarMappedProps {
  contact: ContactModel | null
  idCheck: IdentityCheckModel | null
  status: SectionsStatus
}

export const mapStateToProps = (state: ReduxState): AMLProgressBarMappedProps => ({
  contact: selectCheckListDetailContact(state),
  idCheck: selectCheckListDetailIdCheck(state),
  status: selectCheckListDetailStatus(state),
})

export interface AMLProgressBarMappedActions {
  showModal: (modalType: string) => void
}

export const mapDispatchToProps = (dispatch: Dispatch): AMLProgressBarMappedActions => ({
  showModal: (modalType: string) => dispatch(checklistDetailShowModal(modalType)),
})

export default connect(mapStateToProps, mapDispatchToProps)(AMLProgressBar)
