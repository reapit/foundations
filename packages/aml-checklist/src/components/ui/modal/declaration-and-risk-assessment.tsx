import React from 'react'
import { Input, Button, CameraImageInput, SelectBox, Formik, Form } from '@reapit/elements'
import { connect } from 'react-redux'
import { RISK_ASSESSMENT_TYPE } from '@/constants/appointment-detail'
import { ContactModel } from '@reapit/foundations-ts-definitions'
import styles from '@/styles/pages/checklist-detail.scss?mod'
import { ReduxState } from '@/types/core'
import { checklistDetailShowModal, updateDeclarationAndRisk } from '@/actions/checklist-detail'
import { STEPS } from './modal'
import { Dispatch } from 'redux'
import { validate } from '@/utils/form/declaration-and-risk-assessment'

const optionsRiskAssessmentType = [
  { label: 'Please select...', value: '' },
  { label: RISK_ASSESSMENT_TYPE.SIMPLIFIED, value: RISK_ASSESSMENT_TYPE.SIMPLIFIED },
  { label: RISK_ASSESSMENT_TYPE.NORMAL, value: RISK_ASSESSMENT_TYPE.NORMAL },
  { label: RISK_ASSESSMENT_TYPE.ENHANCED, value: RISK_ASSESSMENT_TYPE.ENHANCED },
]

export const renderForm = ({ onNextHandler, onPrevHandler, isSubmitting }) => ({ values }) => {
  return (
    <Form>
      <div>
        <div>
          <label className="label">Declaration Form</label>
          <CameraImageInput
            labelText="Upload file"
            id="metadata.declarationRisk.declarationForm"
            name="metadata.declarationRisk.declarationForm"
            allowClear={true}
          />
        </div>
        <SelectBox
          labelText="Risk Assessment Type"
          id="metadata.declarationRisk.type"
          name="metadata.declarationRisk.type"
          options={optionsRiskAssessmentType}
          required
        />
        <Input
          type="text"
          labelText="Reason for Type"
          id="metadata.declarationRisk.reason"
          name="metadata.declarationRisk.reason"
          required
        />
        <div>
          <label className="label">Risk Assessment Form</label>
          <CameraImageInput
            labelText="Upload file"
            id="metadata.declarationRisk.riskAssessmentForm"
            name="metadata.declarationRisk.riskAssessmentForm"
            allowClear={true}
            required
            accept="image/*"
          />
        </div>
      </div>
      <div className={styles.footerBtn}>
        <Button loading={isSubmitting} className="mr-2" variant="primary" type="submit">
          Save
        </Button>
        <Button disabled={isSubmitting} className="mr-2" variant="primary" type="button" onClick={onPrevHandler}>
          Previous
        </Button>
        <Button loading={isSubmitting} variant="primary" type="button" onClick={onNextHandler(values)}>
          Finish
        </Button>
      </div>
    </Form>
  )
}

export type DeclarationAndRiskAssessmentProps = StateProps & DispatchProps

export const DeclarationAndRiskAssessment: React.FC<DeclarationAndRiskAssessmentProps> = ({
  contact,
  onNextHandler,
  onPrevHandler,
  onHandleSubmit,
  isSubmitting,
}) => {
  const metadata = contact?.metadata || {
    declarationRisk: {
      type: '',
      reason: '',
    },
  }
  const initialValues = React.useMemo(
    () => ({
      metadata,
    }),
    [contact],
  )
  return (
    <div>
      <Formik initialValues={initialValues} onSubmit={onHandleSubmit} validate={validate}>
        {renderForm({
          onNextHandler,
          onPrevHandler,
          isSubmitting,
        })}
      </Formik>
    </div>
  )
}

export const mapStateToProps = (state: ReduxState): StateProps => {
  return {
    isSubmitting: state?.checklistDetail?.isSubmitting || false,
    contact: state?.checklistDetail?.checklistDetailData?.contact || {},
  }
}

export type StateProps = {
  isSubmitting: boolean
  contact: ContactModel
}

export type DispatchProps = {
  onNextHandler: (values: any) => () => void
  onPrevHandler: () => void
  onHandleSubmit: (values: any) => void
}

export type OwnPropsProps = {
  id: string
}

export const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    onHandleSubmit: values => dispatch(updateDeclarationAndRisk({ contact: values })),
    onNextHandler: (values: any) => () => dispatch(updateDeclarationAndRisk({ contact: values })),
    onPrevHandler: () => dispatch(checklistDetailShowModal(STEPS.ADDRESS_INFORMATION)),
  }
}

export const DeclarationAndRiskAssessmentWithRedux = connect(
  mapStateToProps,
  mapDispatchToProps,
)(DeclarationAndRiskAssessment)

DeclarationAndRiskAssessmentWithRedux.displayName = 'DeclarationAndRiskAssessmentWithReduxWithRedux'

export default DeclarationAndRiskAssessmentWithRedux
