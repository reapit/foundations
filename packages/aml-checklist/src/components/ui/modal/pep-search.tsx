import React from 'react'
import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import { FaRegTimesCircle } from 'react-icons/fa'
import { Input, Button, H4, FlexContainerBasic, Loader, Formik, Form } from '@reapit/elements'
import { ReduxState } from '@/types/core'
import styles from '@/styles/pages/checklist-detail.scss?mod'
import {
  checklistDetailShowModal,
  checklistDetailSearchPep,
  checklistDetailHideModal,
} from '@/actions/checklist-detail'
import { STEPS } from './modal'
import { getPepSearchStatus } from '@/utils/pep-search'
import { ContactModel } from '@reapit/foundations-ts-definitions'

export const renderLoading = () => {
  return (
    <div className={styles.pepSearchLoading}>
      <Loader body={false} />
    </div>
  )
}

export const renderNoResult = (param: string, time: string) => {
  return (
    <FlexContainerBasic className={styles.noResultContainer}>
      <FaRegTimesCircle />
      <div>
        <H4 className={styles.noResultTitle}>No Result Found for &lsquo;{param}&rsquo;.</H4>
        <p>
          Search conducted for &lsquo;{param}&rsquo; on {time}
        </p>
      </div>
    </FlexContainerBasic>
  )
}

// eslint-disable-next-line
export const renderForm = ({ onPrevHandler, onNextHandler, isSubmitting, pepSearchStatus, contact }) => () => {
  const { param, result, time } = pepSearchStatus && contact.id && pepSearchStatus[contact.id]
  return (
    <Form>
      <div className={styles.pepSearchInputContainer}>
        <Input type="text" name="name" id="name" labelText="Enter a name" />
        <Button type="submit" variant="primary" className="mt-2">
          Search
        </Button>
      </div>
      {isSubmitting ? renderLoading() : null}
      {result && result.length === 0 && !isSubmitting ? renderNoResult(param, time) : null}
      <div className="flex justify-end">
        <Button
          disabled={isSubmitting}
          type="button"
          variant="primary"
          className="mr-2"
          dataTest="prev-btn"
          onClick={onPrevHandler}
        >
          Previous
        </Button>
        <Button disabled={isSubmitting} type="button" variant="primary" dataTest="next-btn" onClick={onNextHandler}>
          Finish
        </Button>
      </div>
    </Form>
  )
}

export type PepSearchProps = StateProps & DispatchProps

export const PepSearch: React.FC<PepSearchProps> = ({
  contact,
  handleSubmit,
  onPrevHandler,
  onNextHandler,
  isSubmitting,
}) => {
  const pepSearchStatus = getPepSearchStatus()
  return (
    <div>
      <Formik initialValues={{ name: `${contact.forename} ${contact.surname}` }} onSubmit={handleSubmit}>
        {renderForm({
          onPrevHandler,
          onNextHandler,
          isSubmitting,
          pepSearchStatus,
          contact,
        })}
      </Formik>
    </div>
  )
}

export type StateProps = {
  isSubmitting: boolean
  contact: ContactModel
}

export const mapStateToProps = (state: ReduxState) => {
  return {
    contact: state?.checklistDetail?.checklistDetailData?.contact || {},
    isSubmitting: state?.checklistDetail?.isSubmitting || false,
  }
}

export type DispatchProps = {
  handleSubmit: (values: any) => void
  onNextHandler: () => void
  onPrevHandler: () => void
}

export const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    handleSubmit: values => {
      dispatch(checklistDetailSearchPep(values.name))
    },
    onPrevHandler: () => dispatch(checklistDetailShowModal(STEPS.DECLARATION_RISK_MANAGEMENT)),
    onNextHandler: () => dispatch(checklistDetailHideModal()),
  }
}

const PepSearchWithRedux = connect(mapStateToProps, mapDispatchToProps)(PepSearch)

PepSearchWithRedux.displayName = 'PepSearchWithRedux'

export default PepSearchWithRedux
