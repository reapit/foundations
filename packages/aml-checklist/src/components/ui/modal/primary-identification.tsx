import * as React from 'react'
import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import { ReduxState } from '@/types/core'
import Identification, {
  IDENTIFICATION_FORM_DEFAULT_VALUES,
  IdentityDocumentForm,
} from '@/components/ui/forms/identification'
import { STEPS } from '@/components/ui/modal/modal'
import { checklistDetailShowModal, checklistDetailPrimaryIdUpdateData } from '@/actions/checklist-detail'
import {
  selectCheckListDetailContact,
  selectCheckListDetailPrimaryId,
  selectCheckListDetailIsSubmitting,
} from '@/selectors/checklist-detail'
import { ContactModel } from '@reapit/foundations-ts-definitions'

export type PrimaryIdentiticationProps = DispatchProps & StateProps

export const PrimaryIdentification: React.FC<PrimaryIdentiticationProps> = ({
  contact,
  initFormValues,
  loading,
  updateIdentification,
  onNextHandler,
  onPrevHandler,
}) => (
  <Identification
    loading={loading}
    initFormValues={initFormValues}
    contact={contact}
    onSaveHandler={updateIdentification}
    onNextHandler={onNextHandler}
    onPrevHandler={onPrevHandler}
  />
)

export type StateProps = {
  loading: boolean
  contact: ContactModel | null
  initFormValues: IdentityDocumentForm
}

export const mapStateToProps = (state: ReduxState): StateProps => {
  const isSubmitting = selectCheckListDetailIsSubmitting(state)
  const contact = selectCheckListDetailContact(state)
  const primaryIdDocument = selectCheckListDetailPrimaryId(state)

  let initFormValues = IDENTIFICATION_FORM_DEFAULT_VALUES
  const DEFAULT_TYPE = ''
  if (primaryIdDocument) {
    const { typeId, expiry, details, documentId } = primaryIdDocument

    initFormValues = {
      typeId: typeId || DEFAULT_TYPE,
      expiry: expiry ? new Date(expiry) : '',
      details: details || '',
      documentId: documentId || '',
    } as IdentityDocumentForm
  }

  return {
    loading: isSubmitting,
    contact,
    initFormValues,
  }
}

export type DispatchProps = {
  updateIdentification: (formValues: IdentityDocumentForm) => void
  onPrevHandler: () => void
  onNextHandler: (values: any) => () => void
}

export const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
  updateIdentification: (values: IdentityDocumentForm) =>
    dispatch(checklistDetailPrimaryIdUpdateData({ identityChecks: values })),
  onPrevHandler: () => dispatch(checklistDetailShowModal(STEPS.PROFILE)),
  onNextHandler: (values: any) => () =>
    dispatch(
      dispatch(
        checklistDetailPrimaryIdUpdateData({ nextSection: STEPS.SECONDARY_IDENTIFICATION, identityChecks: values }),
      ),
    ),
})

export default connect(mapStateToProps, mapDispatchToProps)(PrimaryIdentification)
