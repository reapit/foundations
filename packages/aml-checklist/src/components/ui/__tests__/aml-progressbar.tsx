import React from 'react'
import { shallow } from 'enzyme'
import { AMLProgressBar, calculateProgress, mapStateToProps, mapDispatchToProps } from '../aml-progressbar'
import { sectionsStatus } from '@/sagas/__stubs__/status'
import { contact } from '@/sagas/__stubs__/contact'
import { ReduxState } from '@/types/core'
import { STEPS } from '../modal/modal'
import { idCheck } from '@/sagas/__stubs__/id-check'
import { defaultStatus } from '@/constants/section-status'

describe('AMLProgressBar', () => {
  describe('AMLProgressBar', () => {
    it('should match snapshot', () => {
      const mockProps = {
        contact: contact,
        status: sectionsStatus,
        id: 'AYL19000002',
        loginMode: 'WEB',
        showModal: jest.fn(),
        idCheck: {},
      }
      const wrapper = shallow(<AMLProgressBar {...mockProps} />)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('calculateProgress', () => {
    const result = calculateProgress(sectionsStatus)
    const expected = { percentage: 40, completed: 2, total: 5 }
    expect(result).toEqual(expected)
  })

  describe('mapStateToProps', () => {
    it('should run correctly', () => {
      const mockState = {
        checklistDetail: {
          checklistDetailData: {
            contact,
            idCheck,
          },
          status: sectionsStatus,
        },
      } as ReduxState
      const result = mapStateToProps(mockState)
      expect(result).toEqual({
        contact,
        idCheck,
        status: sectionsStatus,
      })
    })
    it('should run correctly', () => {
      const mockState = {} as ReduxState
      const result = mapStateToProps(mockState)
      expect(result).toEqual({
        contact: null,
        idCheck: null,
        status: defaultStatus,
      })
    })
  })

  describe('mapDispatchToProps', () => {
    it('should render correctly', () => {
      const mockDispatch = jest.fn()
      const { showModal } = mapDispatchToProps(mockDispatch)
      showModal(STEPS.PROFILE)
      expect(mockDispatch).toBeCalled()
    })
  })
})
