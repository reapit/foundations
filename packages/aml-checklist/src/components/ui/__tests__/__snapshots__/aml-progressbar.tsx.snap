// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AMLProgressBar AMLProgressBar should match snapshot 1`] = `
<mock-styled.div>
  <Component>
    <Component>
      <Component>
        <Component>
          <Component
            hasMargin={false}
            hasPadding={false}
            isFlex={true}
            isFlexColumn={true}
          >
            <Component
              className="mb-5"
            >
              <Component
                dynamicLinkParams={
                  Object {
                    "appMode": "WEB",
                    "entityCode": "AYL19000001",
                    "entityType": "contacts",
                  }
                }
              >
                Mrs H Phillips
              </Component>
            </Component>
            <Component />
          </Component>
        </Component>
      </Component>
      <Component>
        <Component>
          <Component
            disabled={false}
            onClick={[Function]}
            type="button"
            variant="primary"
          >
            Update Status
          </Component>
        </Component>
      </Component>
    </Component>
    <ProgressBar
      percentage={40}
    />
    <div
      className="has-text-right"
    >
      2
      /
      5
       
      <span>
        Completed
      </span>
    </div>
  </Component>
</mock-styled.div>
`;
